napari/utils/_proxies.py

```diff
@@ -0,0 +1,81 @@
+import re
+import warnings
+from typing import Any, Callable, Generic, TypeVar
+
+import wrapt
+
+from ..utils.translations import trans
+
+_T = TypeVar("_T")
+
+
+class ReadOnlyWrapper(wrapt.ObjectProxy):
+    """
+    Disable item and attribute setting with the exception of  ``__wrapped__``.
+    """
+
+    def __setattr__(self, name, val):
+        if name != '__wrapped__':
+            raise TypeError(
+                trans._(
+                    'cannot set attribute {name}',
+                    deferred=True,
+                    name=name,
+                )
+            )
+
+        super().__setattr__(name, val)
+
+    def __setitem__(self, name, val):
+        raise TypeError(
+            trans._('cannot set item {name}', deferred=True, name=name)
+        )
+
+
+_SUNDER = re.compile('^_[^_]')
+
+
+class PublicOnlyProxy(wrapt.ObjectProxy, Generic[_T]):
+    """Proxy to prevent private attribute and item access, recursively."""
+
+    __wrapped__: _T
+
+    def __getattr__(self, name: str):
+        if name.startswith('_'):
+            warnings.warn(
+                trans._(
+                    "Private attribute access in this context is deprecated and will be unavailable in version 0.4.14",
```

Comment:
```suggestion
                    "Private attribute access in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14",
```

Just wondering if we can improve this warning a little. Thoughts @jni?

Reply:
could do, but we might want to make a new subclass then (or make the message editable by instances).  As is, this class makes no assumptions about the nature or location of its usage

Reply:
oh i did not realize that. hmm message editable by instance might make sense? or even a kwarg for switching between warning and error with custom strings. it makes this a little be more complex - but then widely reusable

Reply:
just accepted this change.  if we need to generalize it later we can